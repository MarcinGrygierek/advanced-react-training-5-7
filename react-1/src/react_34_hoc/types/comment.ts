export interface Comment {
    title: string;
    text: string;
}