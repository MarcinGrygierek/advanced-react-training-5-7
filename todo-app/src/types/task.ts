interface TaskBase {
    id: string;
    name: string;
    description: string;
}

export interface SingleTask extends TaskBase {
    createdAt: Date;
    finishedAt: Date | null;
}

export interface TaskRes extends TaskBase {
    createdAt: string;
    finishedAt: string | null;
}

export interface TaskToCreate {
    name: string;
    description: string;
}

export interface TaskStatusReq {
    finishedAt: Date | null;
}