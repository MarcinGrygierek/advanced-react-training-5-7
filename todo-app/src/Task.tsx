import { differenceInSeconds } from "date-fns";
import { SingleTask } from "./types/task";

interface Props {
    task: SingleTask;
    finishTask: (taskId: string) => void;
    reopenTask: (taskId: string) => void;
}

const Task = ({task, finishTask, reopenTask}: Props) => {

    const calculateDifference = () =>  {
        if(!task.finishedAt) return null;
        return differenceInSeconds(task.finishedAt, task.createdAt);
    }
    

    return (
        <li className="card mt-3">
            <div className="card-body">
                <h2 className="cart-title">{task.name}</h2>
                <p>Created at: {task.createdAt.toLocaleString()}</p>
                {task.finishedAt && <>
                    <p>Finished at: {task.finishedAt.toLocaleString()}</p>
                    <p>Elapsed time <div className="badge bg-primary">{calculateDifference()} s</div></p>
                </>}

                {task.finishedAt 
                ? <button className='btn btn-lg btn-secondary' onClick={() => reopenTask(task.id)}>Reopen</button>  
                : <button className='btn btn-lg btn-success' onClick={() => finishTask(task.id)}>Close</button> }
            </div>
        </li>
    )
}
{/* 
        <div className="card mt-3">
          <div className="card-body">
            <h2 className="cart-title">Closed task</h2>
            <p>Elapsed time <div className="badge bg-primary">00:10</div></p>
          
          </div>
        </div> */}

export { Task };
