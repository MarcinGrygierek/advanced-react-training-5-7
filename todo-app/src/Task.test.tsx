import * as renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import { Task } from './Task';
import { SingleTask } from './types/task';

describe('Task', () => {
    it('Task renders as opened', () => {
        const tree = renderer
        .create(
            <Task
                task={{
                    id: '1',
                    name: 'Test',
                    description: 'Test',
                    createdAt: new Date('7.07.2021, 07:00:00'),
                    finishedAt: null
                }}
                finishTask={() => {}}
                reopenTask={() => {}}
            />
        ).toJSON();
        expect(tree).toMatchSnapshot()
    })

    it('Task renders as closed', () => {
        const tree = renderer
        .create(
            <Task
                task={{
                    id: '1',
                    name: 'Test',
                    description: 'Test',
                    createdAt: new Date('7.07.2021, 07:00:00'),
                    finishedAt: new Date('8.07.2021, 07:00:00'),
                }}
                finishTask={() => {}}
                reopenTask={() => {}}
            />
        ).toJSON();
        expect(tree).toMatchSnapshot()
    })

    test('Task should be closed properly', async () => {
        const task: SingleTask = {
            id: '1',
            name: 'Test',
            description: 'Test',
            createdAt: new Date('7.07.2021, 07:00:00'),
            finishedAt: null
        }
        const { rerender } = render(<Task
            task={task}
            finishTask={() => {
                task.finishedAt = new Date()
            }}
            reopenTask={() => {}}
        />)

        fireEvent.click(screen.getByTestId('close-button'));

        rerender(<Task
            task={task}
            finishTask={() => {}}
            reopenTask={() => {}}
        />)

        expect(await screen.findByTestId('reopen-button')).toBeInTheDocument();
    })
})