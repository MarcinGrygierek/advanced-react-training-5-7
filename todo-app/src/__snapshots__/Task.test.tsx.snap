// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Task renders as closed 1`] = `
<li
  className="card mt-3 accent"
>
  <div
    className="card-body"
  >
    <h2
      className="card-title"
    >
      Test
    </h2>
    <p>
      <small>
        Test
      </small>
    </p>
    <p>
      Created at: 
      7.07.2021, 07:00:00
    </p>
    <p>
      Finished at: 
      7.08.2021, 07:00:00
    </p>
    <div>
      Elapsed time 
      <div
        className="badge bg-primary"
      >
        2678400
         s
      </div>
    </div>
    <button
      className="btn btn-lg btn-secondary"
      data-testid="reopen-button"
      onClick={[Function]}
    >
      Reopen
    </button>
  </div>
</li>
`;

exports[`Task Task renders as opened 1`] = `
<li
  className="card mt-3 accent"
>
  <div
    className="card-body"
  >
    <h2
      className="card-title"
    >
      Test
    </h2>
    <p>
      <small>
        Test
      </small>
    </p>
    <p>
      Created at: 
      7.07.2021, 07:00:00
    </p>
    <button
      className="btn btn-lg btn-success"
      data-testid="close-button"
      onClick={[Function]}
    >
      Close
    </button>
  </div>
</li>
`;
